lib/documents.js

```diff
@@ -2263,12 +2268,184 @@ function writeAllDocumentsImpl(options) {
+      break;
+    case 'hosts':
+      maxRequesters = jobState.jobOptions.concurrentRequests.multiplier *
+          new Set(output.map((forest) => forest.host)).size;
+      break;
+    default:
+      throw new Error('Invalid value for onCompletion.multipleOf. Value must be forests or hosts.');
+  }
+  jobState.startTime = Date.now();
+  for (let i = 0; i < maxRequesters; i++) {
+    jobState.requesterCount++;
+    onWriteAllDocs(jobState,i);
+  }
+}
+
+function onWriteAllDocs(jobState, writerId) {
+
+  const writeBatchArray = [];
+  const batchSize = jobState.jobOptions.batchSize;
+  jobState.lastBatch = false;
+  jobState.retryCount = 2;
+
+  a:
+      for (let i = 0; i < batchSize; i++) {
+        const record = jobState.stream.read();
+
+        if(record !== null) {
+          if (jobState.jobOptions.defaultMetadata.length>0) {
+            writeBatchArray.push(jobState.jobOptions.defaultMetadata);
+          }
+        }
+        else {
+          switch (writeBatchArray.length) {
+            case 0: {
+              jobState.timeElapsed = Date.now() - jobState.startTime;
+              finishWriter(jobState);
+              break a;
+            }
+            default: {
+              jobState.lastBatch = true;
+              break a;
+            }
+          }
+        }
+        writeBatchArray.push(record);
+      }
+
+  if (writeBatchArray.length !== 0) {
+    const batchdef = [];
+    batchdef.push({documents: writeBatchArray});
+    if (jobState.jobOptions.transform.length>0) {
+      batchdef.push({transform:jobState.jobOptions.transform});
+    }
+      writeDocumentsImpl.call(jobState.docInstance, false, batchdef)
+          .result((output) => {
+            onWriteAllDocs(jobState, writerId);
+            jobState.docsWrittenSuccessfully+= writeBatchArray.length;
+              if (jobState.jobOptions.onBatchSuccess) {
+                const progressSoFar = {
+                  docsWrittenSuccessfully: jobState.docsWrittenSuccessfully,
+                  docsFailedToBeWritten: jobState.docsFailedToBeWritten,
+                  timeElapsed: (Date.now()-jobState.startTime)
+                };
+                const documents = writeBatchArray;
+                jobState.jobOptions.onBatchSuccess(progressSoFar, documents);
+              }
+            if(jobState.lastBatch){
+              jobState.timeElapsed = Date.now() - jobState.startTime;
+              finishWriter(jobState);
+            }
+          })
+          .catch((err) => {
+              if (jobState.jobOptions.onBatchError) {
+                const progressSoFar = {
+                  docsWrittenSuccessfully: jobState.docsWrittenSuccessfully,
+                  docsFailedToBeWritten: writeBatchArray.length,
+                  timeElapsed: (Date.now()-jobState.startTime)
+                };
+                const documents = writeBatchArray;
+                jobState.jobOptions.onBatchError(progressSoFar, documents, err);
```

Comment:
A future iteration will need to process the return value from onBatchError() as described in the story, but that iteration can occur after this iteration is merged.